### PR TITLE
[RyuJIT/ARM32] Add the optimization case on CreateDictionaryLookupHelper

### PR DESCRIPTION
**Pull request by sjsinju**
#13653
#13952

I implemented CreateDictionaryLookupHelper only via helper (#13733) before. For the optimization case, This PR implement additional codes for found data on dictionary.

### CODE REVIEW

**Comment by sjsinju** [comment]
I implemented the codes similar with ARM64 but it makes segmentation fault on running below test with R2R image.
```TC/readytorun/tests/generics/generics.ni.exe```

Asm codes were made well.
```
   0x6f9a6c3c:	mov	r3, r0
   0x6f9a6c3e:	ldr.w	r0, [r0, #36]	; 0x24
   0x6f9a6c42:	ldr.w	r0, [r0]                              << Segmentation Fault
   0x6f9a6c46:	ldr.w	r0, [r0, #4]
   0x6f9a6c4a:	cbz	r0, 0x6f9a6c4e
   0x6f9a6c4c:	mov	pc, lr
   0x6f9a6c4e:	mov	r0, r3
   0x6f9a6c50:	movw	r1, #27696	; 0x6c30
   0x6f9a6c54:	movt	r1, #28570	; 0x6f9a
   0x6f9a6c58:	movw	r12, #38221	; 0x954d
   0x6f9a6c5c:	movt	r12, #30293	; 0x7655
   0x6f9a6c60:	bx	r12
```
but when the ```ldr.w	r0, [r0]``` is run, the fault is occurred.
I figured out the loaded value from offset 36 is the 16 not address
```
0x6f9a6c42 in ?? ()
(gdb) i r
r0             0x10	16
r1             0x7dd8f800	2111371264
r2             0x1	1
r3             0x75a198ec	1973524716
r4             0x766a90c9	1986695369
r5             0x72547ae8	1918139112
r6             0x725479b0	1918138800
r7             0x735010bc	1934627004
r8             0x9	9
r9             0x7dd8f800	2111371264
r10            0x76fff000	1996484608
r11            0x7effe080	2130698368
r12            0x6f9a6c3d	1872391229
sp             0x7effe070	0x7effe070
lr             0x715a3ce9	1901739241
pc             0x6f9a6c42	0x6f9a6c42
cpsr           0x600d0030	1611464752
(gdb) ni

Thread 1 "corerun" received signal SIGSEGV, Segmentation fault.
0x6f9a6c42 in ?? ()
```

**Comment by sjsinju** [comment]
@gbalykov I think it is related with #12168. 16 looks like relative value of the address.
Can you help me solve this problem?

**Comment by gbalykov** [comment]
`CORINFO_RUNTIME_LOOKUP` structure has two fields:

```cpp
    bool                indirectFirstOffset;
    bool                indirectSecondOffset;
```

You need to check for the `pLookup->offsets[1]` and the `pLookup->offsets[2]`. You can see `Compiler::impRuntimeLookupToTree` for details, it performs checks in the same manner, as you will need to implement in your code.

https://github.com/dotnet/coreclr/blob/f5b7f8ba68688382e57999e33ae39f46ddadd47a/src/jit/importer.cpp#L1973

For example,

```cpp
    if (pLookup->indirectFirstOffset)
    {
        // generate code with indirection
    }
    else
    {
        // generate code without indirection (your current code)
    }

```

`indirectFirstOffset` and `indirectSecondOffset` basically mean that offset is relative and instead of 

```asm
    mov r0, [r0 + offset]
```

you need to generate something like

```asm
    mov r1, [r0 + offset]
    add r0, r1
```

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT arm Cross Checked Build and Test  please

**Comment by sjsinju** [comment]
@gbalykov Thanks for your helpful comments! Let me check what you said.

**Comment by hqueue** [comment]
@dotnet-bot please test Tizen armel Cross Release Build 
@dotnet-bot please test Tizen armel Cross Debug Build

**Comment by hqueue** [comment]
@sjsinju I've update your PR with some fixes in a conservative way and it worked :)
Please take a look at https://github.com/hqueue/coreclr/commits/arm/r2r/add_dic_optimized_helper

It passes `generics.ni.exe`.

**Comment by sjsinju** [comment]
@hqueue  Thank you very much!! I've checked your code works well. I will make another PR for instruction optimization based your PR.

**Comment by sjsinju** [comment]
/cc @dotnet/arm32-contrib @dotnet/arm64-contrib 
@gbalykov @hqueue It works well now. Could you please check this?

**Comment by sjsinju** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test
@dotnet-bot testUbuntu x64 Checked Build and Test

**Comment by sjsinju** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

File: src/vm/arm/stubs.cpp [hunk 1]
```diff
@@ -3782,16 +3782,122 @@ PCODE DynamicHelpers::CreateDictionaryLookupHelper(LoaderAllocator * pAllocator,
...
+
+                    // add r0, r2
+                    *(WORD *)p = 0x4410;
+                    p += 2;
+                }
+                else
+                {
+                    // add r0, r0, <offset>
+                   *(WORD *)(p + 0) = 0xF100;
+                   *(WORD *)(p + 2) = pLookup->offsets[i];
```
Comment (hqueue) [inline src/vm/arm/stubs.cpp]: I'm afraid we have to encode immediate here. According to ARM manual, thumb instruction for `add r0, r0, #imm12` has different encoding. Please check the encoding.

Comment (hqueue) [reply src/vm/arm/stubs.cpp]: Also please use bitmasks and type casting as in below code for `// ldr r0, [r0 + offset]`

Comment (hqueue) [reply src/vm/arm/stubs.cpp]: In addition, instead of using 4 bytes `add r0, r0, #imm12`, we can use 2 bytes `add r0, #imm8`.
What do you think of it?

Comment (sjsinju) [reply src/vm/arm/stubs.cpp]: @hqueue OK. I've just copied and pasted, So it made some mistake. Thank you.

**Comment by sjsinju** [comment]
@jashook  @jkotas Can you check and merge this??

**Comment by jkotas** [comment]
Could you please reenable the tests disabled against the bug (revert #13954) in the PR?

**Comment by sjsinju** [comment]
@jkotas The tests what you say will be enabled by additional commit of this PR.

**Pull request merged**